src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs

```diff
@@ -0,0 +1,106 @@
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+
+namespace MongoDB.Driver.Core.Operations
+{
+    internal static class RetryabilityHelper
+    {
+        // private static fields
+        private static readonly HashSet<ServerErrorCode> __notResumableChangeStreamErrorCodes;
+        private static readonly HashSet<Type> __resumableChangeStreamExceptions;
+        private static readonly HashSet<ServerErrorCode> __resumableChangeStreamErrorCodes;
+        private static readonly HashSet<Type> __retryableWriteExceptions;
+        private static readonly HashSet<ServerErrorCode> __retryableWriteErrorCodes;
+
+        // static constructor
+        static RetryabilityHelper()
+        {
+            var resumableAndRetryableExceptions = new HashSet<Type>()
+            {
+                typeof(MongoConnectionException),
+                typeof(MongoNotPrimaryException),
+                typeof(MongoNodeIsRecoveringException)
+            };
+
+            __resumableChangeStreamExceptions = new HashSet<Type>(resumableAndRetryableExceptions)
+            {
+                typeof(MongoCursorNotFoundException)
+            };
+
+            __retryableWriteExceptions = new HashSet<Type>(resumableAndRetryableExceptions)
+            {
+            };
+
+            var resumableAndRetryableErrorCodes = new HashSet<ServerErrorCode>
+            {
+                ServerErrorCode.HostNotFound,
+                ServerErrorCode.HostUnreachable,
+                ServerErrorCode.NetworkTimeout,
+                ServerErrorCode.SocketException
+            };
+
+            __resumableChangeStreamErrorCodes = new HashSet<ServerErrorCode>(resumableAndRetryableErrorCodes)
+            {
+                ServerErrorCode.ElectionInProgress,
+                ServerErrorCode.ExceededTimeLimit,
+                ServerErrorCode.RetryChangeStream,
+            };
+
+            __retryableWriteErrorCodes = new HashSet<ServerErrorCode>(resumableAndRetryableErrorCodes)
+            {
+                ServerErrorCode.WriteConcernFailed
+            };
+
+            __notResumableChangeStreamErrorCodes = new HashSet<ServerErrorCode>()
+            {
+                ServerErrorCode.CappedPositionLost,
+                ServerErrorCode.CursorKilled,
+                ServerErrorCode.Interrupted
+            };
+        }
+
+        // public static methods
+        public static bool IsResumableChangeStreamException(Exception exception)
+        {
+            var commandException = exception as MongoCommandException;
+            if (commandException != null)
+            {
+                var code = (ServerErrorCode)commandException.Code;
+                return __resumableChangeStreamErrorCodes.Contains(code) || !__notResumableChangeStreamErrorCodes.Contains(code);
```

Comment:
This test is suspect but that's the way the spec is written. I've pointed this out to Kevin and he might update the spec.

The reason it is suspect is because the two Contains tests are redundant. If the code is in __resumableChangeStreamErrorCodes it is automatically NOT in __notResumableChangeStreamErrorCodes.

Reply:
I'm not clear on why you don't just get rid of __resumableChangeStreamErrorCodes and simplify this conditional. All the tests will still pass without it, right?

Reply:
I agree.

I wrote it this way because that what the spec said to do.

But I think the spec needs to be revised. Kevin is looking into it.

Reply:
OK.  LGTM on this, but feel free to hold off a bit while you get clarification.